worker/src/queue.rs

```diff
@@ -179,15 +342,343 @@ impl AsRef<JsValue> for Queue {
+            Self::Json => "json",
+            Self::Text => "text",
+            Self::V8 => "v8",
+        })
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct QueueSendOptions {
+    content_type: Option<QueueContentType>,
+    delay_seconds: Option<u32>,
+}
+
+pub struct MessageBuilder<T> {
+    message: T,
+    delay_seconds: Option<u32>,
+    content_type: QueueContentType,
+}
+
+impl<T: Serialize> MessageBuilder<T> {
+    /// Creates a new message builder. The message must be `serializable`.
+    pub fn message(message: T) -> Self {
+        Self {
+            message,
+            delay_seconds: None,
+            content_type: QueueContentType::Json,
+        }
+    }
+
+    #[must_use]
+    /// The number of seconds to delay a message for within the queue, before it can be delivered to a consumer
+    pub fn with_delay_seconds(mut self, delay_seconds: u32) -> Self {
+        self.delay_seconds = Some(delay_seconds);
+        self
+    }
+
+    #[must_use]
+    /// The content type of the message.
+    /// Default is `QueueContentType::Json`.
+    pub fn with_content_type(mut self, content_type: QueueContentType) -> Self {
+        self.content_type = content_type;
+        self
+    }
+
+    /// Build the message.
+    pub fn build(self) -> SendMessage<T> {
+        SendMessage {
+            message: self.message,
+            options: Some(QueueSendOptions {
+                content_type: Some(self.content_type),
+                delay_seconds: self.delay_seconds,
+            }),
+        }
+    }
+}
+
+pub struct RawMessageBuilder {
+    message: JsValue,
+    delay_seconds: Option<u32>,
+}
+
+impl RawMessageBuilder {
+    /// Creates a new raw message builder. The message must be a `JsValue`.
+    pub fn message(message: JsValue) -> Self {
+        Self {
+            message,
+            delay_seconds: None,
+        }
+    }
+
+    #[must_use]
+    /// The number of seconds to delay a message for within the queue, before it can be delivered to a consumer
+    pub fn with_delay_seconds(mut self, delay_seconds: u32) -> Self {
+        self.delay_seconds = Some(delay_seconds);
+        self
+    }
+
+    /// Build the message with a content type.
+    pub fn build_with_content_type(self, content_type: QueueContentType) -> SendMessage<JsValue> {
+        SendMessage {
+            message: self.message,
+            options: Some(QueueSendOptions {
+                content_type: Some(content_type),
+                delay_seconds: self.delay_seconds,
+            }),
+        }
+    }
+}
+
+pub struct SendMessage<T> {
+    message: T,
+    options: Option<QueueSendOptions>,
+}
+
+impl<T: Serialize> SendMessage<T> {
+    pub fn into_raw_message(self) -> Result<SendMessage<JsValue>> {
+        Ok(SendMessage {
+            message: serde_wasm_bindgen::to_value(&self.message)?,
+            options: self.options,
+        })
+    }
+}
+
+impl<T: Serialize> From<T> for SendMessage<T> {
```

Comment:
Could / Should you also implement `From<JsValue>` for easily sending raw messages? Or would it be impossible to discern V8 types?

Reply:
> Could you also implement From `<JsValue>`

Rust won't allow it as the upstream crate may add a new impl of trait `serde::Serialize` for type `wasm_bindgen::JsValue` in future versions

> Should you also implement From<JsValue> for easily sending raw messages? Or would it be impossible to discern V8 types?

I think it's the same issue as in https://github.com/cloudflare/workers-rs/pull/335#discussion_r1543841021 , AFAIK we can't discern the type (I only had a quick look at the js_sys, so I could be missing something), so we would have to decide on a default content type for raw messages.